File: tools-perf-hypack-2013/tools-perf-codes/c/Tools_Pi_Idleness.c
/*
*********************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

  Example 4 : (Tools_Pi_Idleness.c)	

  Objective : To compute the value of PI by numerical integration.
              MPI Point-to-Point communication and collective communication 
              library calls are used to show the idleness in a program.

	      This example demonstrates the use of 
	      MPI_Init
              MPI_Comm_rank
              MPI_Comm_size
              MPI_Bcast
              MPI_Send
              MPI_Recv
              MPI_Finalize

  Input     : The number of intervals.

  Output    : The calculated value of PI.

  Necessary Condition : Number of Processes should be less 
                        than or equal to 8.

 Created             : August-2013

 E-mail              : hpcfte@cdac.in     


*********************************************************************
*/

#include "mpi.h"
#include <stdio.h>
#include <math.h>


double f(double);

double f(double a)
{
    return (4.0 / (1.0 + a*a));
}

int main(int argc,char *argv[])
{
    int done = 0, NoofIntervals, MyRank, Numprocs, i;
    double PI25DT = 3.141592653589793238462643;
    double mypi=0, pi, h, sum, x;
    double startwtime = 0.0, endwtime;
    int  namelen;
    char processor_name[MPI_MAX_PROCESSOR_NAME];
    MPI_Status status;
    
    /* Initializing MPI CALLS */
    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&Numprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&MyRank);


    NoofIntervals = 0;
        if (MyRank == 0)
        {
	    if (NoofIntervals==0) NoofIntervals=10000; else NoofIntervals=0;

        }
       /* broadcasting no of intervals to all the processors */
        MPI_Bcast(&NoofIntervals, 1, MPI_INT, 0, MPI_COMM_WORLD);
        if (NoofIntervals == 0)
            done = 1;
        else
        {
         if(MyRank==0)
           {
            h   = 1.0 / (double) NoofIntervals;
            sum = 0.0;
             for (i = MyRank + 1; i <= NoofIntervals; i += Numprocs)
             {
              x = h * ((double)i - 0.5);
             sum += f(x);
             }
              mypi= h * sum;  
            MPI_Send(&mypi,1,MPI_DOUBLE,MyRank+1,0,MPI_COMM_WORLD);
          }
          else
           { 
             if(MyRank< Numprocs-1)
             { 
            MPI_Recv(&mypi,1,MPI_DOUBLE,MyRank-1,0,MPI_COMM_WORLD,&status);
            h   = 1.0 / (double) NoofIntervals;
            sum = 0.0;
            for (i = MyRank + 1; i <= NoofIntervals; i += Numprocs)

            {
                x = h * ((double)i - 0.5);
                sum += f(x);
            }
            mypi+= h * sum;
            MPI_Send(&mypi,1,MPI_DOUBLE,MyRank+1,0,MPI_COMM_WORLD); 
            }

            if (MyRank == Numprocs-1)
	    {
              MPI_Recv(&mypi,1,MPI_DOUBLE,MyRank-1,0,MPI_COMM_WORLD,&status);
                h   = 1.0 / (double) NoofIntervals;
                sum = 0.0;
                for (i = MyRank + 1; i <= NoofIntervals; i += Numprocs)
                {
                   x = h * ((double)i - 0.5);
                   sum += f(x);
                }
                mypi+= h * sum;
           /* displaying the final result */
                printf("pi is approximately %.16f, Error is %.16f\n",
                       mypi, fabs(mypi - PI25DT));
		fflush( stdout );
	    }
           } 
        
    }
    MPI_Finalize();
}

File: tools-perf-hypack-2013/tools-perf-codes/Fortran/Tools_Pi_Idleness.f
c
c********************************************************************
c
c		C-DAC Tech Workshop : hyPACK-2013
c                           October 15-18, 2013
c
c                   Example 4 (Tools_Pi_Idleness.f)
c
c
c    Objective           : To compute the value of PI by numerical integration.
c                          MPI Point-to-Point Communication and collective
c                          communication libraries are used to show idleness in c program.
c
c                          This example demonstrates the use of
c                          MPI_Init
c                          MPI_Comm_rank
c                          MPI_Comm_size
c                          MPI_Bcast
c                          MPI_Send
c                          MPI_Recv
c                          MPI_Finalize
c
c    Input               : The number of intervals
c
c    Output              : The calculated value of PI.
c
c    Necessary Condition : Number of Processes should be
c                          less than or equal to 8.
c
c   Created             : August-2013
c
c   E-mail              : hpcfte@cdac.in     
c
c
c*********************************************************************      
      program main

      include 'mpif.h'

      double precision  PI25DT
      parameter        (PI25DT = 3.141592653589793238462643d0)

      double precision  mypi, h, sum, x, f, a
      integer NoofIntervals, myid, numprocs, Root
		integer status(MPI_STATUS_SIZE)

C     function to integrate

      f(a) = 4.d0 / (1.d0 + a*a)

      call MPI_INIT( ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, myid, ierr )
      call MPI_COMM_SIZE( MPI_COMM_WORLD, numprocs, ierr )

      Root = 0
      if ( myid .eq. Root ) then
      NoofIntervals = 10 
      end if
      call MPI_BCAST(NoofIntervals,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)
       if ( myid .eq. Root) then
          h = 1.0d0/NoofIntervals
          sum  = 0.0d0
          do 20 i = myid+1, NoofIntervals, numprocs
             x = h * (dble(i) - 0.5d0)
             sum = sum + f(x)
 20       continue
         mypi = h * sum
 	        call MPI_Send(mypi, 1, MPI_DOUBLE_PRECISION, myid+1, 
     $			  0, MPI_COMM_WORLD, ierror)
            endif
		 if( myid .lt. numprocs-1) then
          call MPI_Recv(mypi, 1, MPI_DOUBLE_PRECISION, myid-1,0, 
     $		 MPI_COMM_WORLD, status, ierror)
          h = 1.0d0/NoofIntervals
          sum  = 0.0d0
          do 40 i = myid+1, NoofIntervals, numprocs
             x = h * (dble(i) - 0.5d0)
             sum = sum + f(x)
 40       continue
         mypi = mypi +  h * sum
 	        call MPI_Send(mypi, 1, MPI_DOUBLE_PRECISION, myid+1, 
     $			  0, MPI_COMM_WORLD, ierror)
         endif
         if(myid .eq. numprocs-1) then
          call MPI_Recv(mypi, 1, MPI_DOUBLE_PRECISION, myid-1,0,
     $           MPI_COMM_WORLD, status, ierror)
          h = 1.0d0/NoofIntervals
          sum  = 0.0d0
          do 50 i = myid+1, NoofIntervals, numprocs
             x = h * (dble(i) - 0.5d0)
             sum = sum + f(x)
 50       continue
          mypi = mypi + h * sum 
         write(6, 97) mypi, abs(mypi - PI25DT)
 97      format('  pi is approximately: ', F18.16,
     +          '  Error is: ', F18.16)
          endif
       if ( NoofIntervals .eq. 0 ) then  
       goto 30
       end if
 30   call MPI_FINALIZE(ierr)
      stop
      end





